File: 가원/week1/388351_유연근무제.py
import sys


def solution(schedules, timelogs, startday):
    '''
    params
    - schedules: 1차원 배열 형태의 각 직원이 원하는 출근 희망 시간
    - timelogs: 직원들이 일주일 동안 출근한 시각을 담은 2차원 정수 배열 (실제 직원들 출근시간)
    - startday: 이벤트 시작한 요일을 의미하는 정수
    
    우리가 원하는 결과값은 일주일동안 직원들이 자기가 원하는 출근 희망 시간 지각안하고
    재시간에 출근한 직원의 수
    '''
    answer = 0

    for i in range(len(schedules)):
        goal_time = schedules[i]

        # 지각 허용 시간 계산(희망 시간 + 10분)
        h, m = divmod(goal_time, 100)
        m += 10

        if m >= 60:
            h += 1
            m -= 60
          
        limit_time = h * 100 + m   # 지각 허용시간 완성 -> 이제부터 limit_time  이 마지노선 시간

        # 상품받을 자격 적격 심사
        is_eligible = True
        
        # 이제 지각 마지노선 시간과 실제 직원의 출근시간 비교
        for j in range(7):
            # 현재 요일 계산 (1: 월, 2: 화, 3: 수,...., 7: 일) 
            current_day = (startday + j - 1) % 7 + 1

            # 주말(6, 7)은 패스
            if current_day >= 6:
                continue

            # 평일인데 지각 있으면 탈락
            if timelogs[i][j] > limit_time:
                is_eligible = False
                break

        # 만약에 여기까지 통과했다면 이제 여기 아래부터는 지각을 하지 않은 직원
        if is_eligible:
            answer += 1

    return answer


if __name__ == "__main__":
    print("Hello")

File: 민혁/week1/388351_유연근무제.java
package week1;

/*
각 직원의 출근희망시간과 실제 출근시간을 비교하는 문제로 이해

1. 직원의 출근 희망시간과 실제 출근 시간을 비교

    n(=schedules.length)만큼 반복 -> (각각의 직원의 출근시간을 확인)
        -timelogs만큼 반복 -> (각각의 직원이 출근시간을 매일 잘 지켰는지 확인)
            --주말인지 체크(주말이라면 확인 예외)
            --직원들이 설정한 출근시간(schedules)에서 +10분 안에 드는지 체크
            --평일 모두 조건을 만족하면 answer++

2. return answer

시간 복잡도 O(n)
 */

public class Main {
    public int solution(int[] schedules, int[][] timelogs, int startday) {
        int answer = 0;

        for (int i = 0; i < schedules.length; i++) { // 직원 수만큼 확인
            boolean commute = true;
            for (int j = 0; j < timelogs[i].length; j++) { //개별 직원이 매일 잘 출근했는지 확인

                //시간 비교를 위해 분으로 변환 (이유) 설정이750인 경우에는 10분 뒤가 800으로 넘어가 +10 비교 불가
                int st = schedules[i] / 100 * 60 + schedules[i] % 100; //schedules 분 변환
                int tt = timelogs[i][j] / 100 * 60 + timelogs[i][j] % 100; // timelogs 분 변환

                //if 주말 체크
                if(j == ((6 - startday) + 7) % 7 || j == ((7 - startday) + 7) % 7) {
                    // (6 - startday) 가 음수로 넘어가는 것을 방지하기 위헤 +7 을 한 뒤 % 7
                    continue;
                }

                // 평일 해당 요일에 출근을 잘했는가?
                if(!(tt <= st + 10)) {
                    commute = false; // 시간 안에 못 왔으면 false로 바꿔줌
                    break; // 하루라도 못지켰으면 그 인원은 out, 반복문 탈출
                }
            }
            // 만약 평일 모두 출근을 잘했다면 commute == true   answer++
            if(commute) {
                answer++;
            }
        }

        return answer;
    }
}
